Added ability to skip generating boilerplate code and runline

gennadiycivil opened the pull request:
New command line flag to fix #47

mbrukman commented on autogen:
```diff
@@ -385,8 +395,12 @@ function autogenForFile() {
+
+      if [ ${ADD_PYTHON_CODE} -eq 1 ]; then
+        echo "#!/usr/bin/python"
+        echo "#"
```
This line should be part of the license block; otherwise, let's just add blank line here?

gennadiycivil replied:
Not sure what you mean, the "#!/usr/bin/python" line is not a part of license... I am probably not understanding you, could you please elaborate

mbrukman replied:
Sorry, I misread the code, so please ignore my earlier comment.

Given this, though, are you suggesting that `#!/usr/bin/python` is also extra if `--no-code` is specified? Is the idea that non-root `*.py` files don't need a `#!/usr/bin/python` line at the top, just the license block?

What about the use case of a root Python file, where you want the `#!/usr/bin/python` line so that it's executable, but without the boilerplate code gen?

mbrukman replied:
Maybe the `#!/usr/bin/python` (or any other line) should only be eliminated if another flag, e.g., `--no-runline` is specified?

I am just suggesting that `--no-code` only removes the code _below_ the license block, and everything above the license block stays as-is. Let's discuss removing the runline in a separate issue / PR. Thoughts?

gennadiycivil replied:
My usage of autogen has largely been driven by a need to add OSS licenses to a large project . I found it very unusual that the tool is also trying to add actual code. ( I consider the #!/usr/bin/python code :-) ) 

So I tell the tool "go to these files, and add license in-place" and the tool does this... but also oh by the way let me add a #!/usr/bin/python line and unit test code -- which was already present in some cases and totally broke the build in others :-1:  

Is the idea that non-root *.py files don't need a #!/usr/bin/python line at the top, just the license block? - Yes, exactly , dont mess with this because as an engineer I will know what to do, in other words, just do what I want.

I realize that there may be other cases and yes,--no-runline makes sense, I ll try to add.

mbrukman replied:
I've been thinking about this a bit and realized that if there's already a `#!/usr/bin/python` at the top of the file, then we don't have a good solution here: either we add a second `#!/usr/bin/python` or if we provide the `--no-runline` then the license header will be at the top, above the existing runline, which is still invalid.

How would you fix existing files with a runline to just add a license block, but _after_ the runline? Not suggesting you should extend your current PR to handle this, just a question.

gennadiycivil replied:
I belived in SOLID principles - the "S" applies here, single responsibility . This tool is designed to help adding boilerplate content to files, it is really not responsible for fixing content that is already there.  So... if the file includes ! line... there must have been a reason to include it, this tool should not do more than it is designed to do...

mbrukman commented on autogen:
```diff
@@ -161,6 +161,11 @@ Options:
                            add a top-level documentation comment after the
                            license header (default: enabled)
 
+  --no-pycode, --no-pyc
```
I think you make a good point overall, but adding a separate flag per-language seems overkill, because it's not possible to generalize it in a config or editor file, for example.

How about a more generic `--no-code` flag that disables all codegen for all languages instead of a Python-specific one?

gennadiycivil replied:
Sound great to me, agreed

mbrukman commented on README.md:
```diff
@@ -53,6 +53,10 @@ To get a list of supported licenses, or to see the full set of flags, run
+The flag `--no-pycode` skips adding extra python boilerplate code to
+python test and python source files:
+
+
```
Extra blank line here.

mbrukman commented on autogen:
```diff
@@ -221,6 +226,11 @@ while :; do
       TOP_LEVEL_COMMENT=0
       ;;
 
+    --no-pycode | --no-pyc)
```
We should also add some tests for new functionality to ensure it does what we think / expect it to do.

gennadiycivil commented:
Please take a look at new changes. I just implemtented python for now. 
Also I added tests but in seemed crazy to add tests for all flag combinations so I added py-no-tls-no-run-line-no-code tests.
if you OK these I will add other languages. Let me know.

mbrukman commented on README.md:
```diff
@@ -53,6 +53,9 @@ To get a list of supported licenses, or to see the full set of flags, run
 File type or language is determined based on the full filename or extension, as
 appropriate. See [`autogen`](autogen) for a list of recognized file types.
 
+The flag `--no-pycode` skips adding extra python boilerplate code to
```
`s/--no-pycode/--no-code/` and also make this statement more generic?

mbrukman replied:
Please update this line.

mbrukman commented on autogen:
```diff
@@ -161,6 +162,13 @@ Options:
+  --no-runline, --no-rnl
+                           Skip adding runline code
+                           to the source files (default: false)
+  --no-code, --no-cod
```
No need to abbreviate "code" as "cod" :-)

Not every flag needs a short-form if it's already short.

mbrukman commented on autogen:
```diff
@@ -161,6 +162,13 @@ Options:
                            add a top-level documentation comment after the
                            license header (default: enabled)
 
+  --no-runline, --no-rnl
```
I would suggest dropping the `--no-rnl` form: it is not a easily-recognized/remembered abbreviation, it's not a first-letter-of-each-of-the-words, so it doesn't add much. If it's ever used in a script, it won't be clear to anyone what it means without looking at the docs, so it's not self-documenting.

I'll admit the `--no-tlc` was a bit of a fun pun for me as much as it was a first-letter-of-each-of-the-words shortening of `--no-top-level-comment`. :-)

gennadiycivil commented:
I added changes for other languages, another unit test, and addressed your comments.
Please take a look, should be close

mbrukman commented on autogen:
```diff
@@ -366,27 +382,39 @@ function autogenForFile() {
     test_*.py | *_test.py)
-      echo "#!/usr/bin/python"
-      echo "#"
+
```
Two extra blank lines here?

mbrukman commented on tests/testdata/py-no-tls-no-run-line-no-code.in:
```diff
@@ -0,0 +1 @@
+--no-tlc --no-code --no-runline file.py
```
The name of this file should be `py-no-tlc` (not tls) `-no-runline-no-code.in` (i.e., mimic the flags on the command line). Similarly for the *.out and *.err files, of course.

mbrukman commented on autogen:
```diff
@@ -366,27 +382,39 @@ function autogenForFile() {
       # "php script.php" .
       #
       # Note: PHP accepts C, C++, and shell-style comments.
-      echo "<?php"
```
I'm not sure if it's valid for a PHP script to have no `<?php` header, will it still parse as a PHP file without it? This is not technically a runline, as there's no `#!/path/to/binary` it's just a marker identifying it as a PHP file.

gennadiycivil replied:
It has been years since I touched PHP... honestly I don't remember. Should I remove it ?

gennadiycivil replied:
I think I should remove it and just leave PHP alone, the NO CODE below also depends on this "> closing comment" so this would break with certain flag combo

mbrukman commented on autogen:
```diff
@@ -366,27 +382,39 @@ function autogenForFile() {
+      if [ ${ADD_RUNLINE} -eq 1 ]; then
+        echo
+        # E_STRICT was added in PHP 5.0 and became included in E_ALL in PHP 6.0 .
+        echo "error_reporting(E_ALL | E_STRICT);"
```
Maybe this is a no-code feature rather than a no-runline?

gennadiycivil replied:
Agreed

mbrukman commented on autogen:
```diff
@@ -161,6 +162,13 @@ Options:
                            license header (default: enabled)
 
+  --no-runline
+                           Skip adding runline code
```
Move the docs one line higher; they're only one line lower when the flag is too long and would overrun this text, which it doesn't in this case.

mbrukman commented on autogen:
```diff
@@ -161,6 +162,13 @@ Options:
+                           Skip adding runline code
+                           to the source files (default: false)
+  --no-code
+                           Skip adding extra bootstrap code
```
Move the docs one line higher; they're only one line lower when the flag is too long and would overrun this text, which it doesn't in this case.

mbrukman commented on README.md:
```diff
@@ -53,6 +53,9 @@ To get a list of supported licenses, or to see the full set of flags, run
 appropriate. See [`autogen`](autogen) for a list of recognized file types.
 
+The flag `--no-pycode` skips adding extra python boilerplate code to
+python test and python source files:
```
Please update this now that it's more generic than previously.

gennadiycivil commented:
PTAL

gennadiycivil commented:
ptal

gennadiycivil commented:
ping

gennadiycivil commented on autogen:
```diff
@@ -434,6 +459,8 @@ EOF
       fi
+      if [ ${ADD_CODE} -eq 1 ]; then
+
       cat <<EOF
```
is this raw content? If I add indent it would break Python compilation ... no?

mbrukman replied:
Sorry, I just meant to indent the`cat`command since it's now 1 level below the `if`, not the Python code block.

gennadiycivil replied:
Done

gennadiycivil commented:
Please take a look

gennadiycivil commented:
ptal

mbrukman commented on autogen:
```diff
@@ -425,7 +425,7 @@ EOF
         fi
         # Add basic bootstrap code.
-        cat <<EOF
+      cat <<EOF
```
This was correct previously: the `cat <<EOF` should be indented here, since it's part of the higher-level `if` statement, and it will fit under the `# Add basic bootstrap code.` comment. Why remove the indent here?

mbrukman commented on autogen:
```diff
@@ -434,6 +459,7 @@ EOF
 EOF
       fi
+      if [ ${ADD_CODE} -eq 1 ]; then
       cat <<EOF
```
This line needs to be indented since it's now under a new `if` statement.

mbrukman commented:
I will squash and merge; the [tests passed on Travis on Linux](https://travis-ci.org/mbrukman/autogen/builds/272656045) but have not even started on macOS, so skipping those.

mbrukman commented:
@gennadiycivil — thank you for your contribution and for your patience with this review!

BTW, since you used your `master` branch, you won't be able to merge this cleanly; please consider deleting your fork on GitHub and creating a new fork. For future PRs, please create a new feature branch for every PR and keep the `master` simply for syncing with upstream (i.e., this repo).

gennadiycivil commented:
@mbrukman thank you. I deleted my fork and re-created new one just in case.
